vec/src/main.rs

```diff
@@ -0,0 +1,365 @@
+                        let (mut i, order): (Vec<_>, Vec<_>) = order.into_iter().unzip();
+
+                        // we'll populate this with the sorted indices first
+                        let mut indices: Vec<usize> = vec![0; i.len()];
+
+                        i.insert(0, 0);
+                        let diffs = i.windows(2).map(|w| w[1] - w[0]);
+
+                        let mut last = iter.nth(0).expect("exhausted iterator");
+                        for (i, diff) in diffs.enumerate() {
+                            if diff > 0 {
+                                last = iter.nth(diff - 1).expect("exhausted iterator");
+                            }
+                            indices[order[i]] = last;
+                        }
+
+                        // and finally we convert our new indices into an iterator
+                        iter = Box::new(indices.into_iter())
+                    }
+                },
+                Subset::Range(range) => {
+                    iter = Box::new(
+                        iter.skip(range.start)
+                            .enumerate()
+                            .take_while(move |(i, _)| i < &(range.end - range.start))
+                            .map(|(_, v)| v)
+                    ) as Self::IntoIter
+                },
+            }
+        }
+
+        iter
+    }
+}
+
+impl<T: Atomic> From<Vec<T>> for Vector<T> {
+    fn from(value: Vec<T>) -> Self {
+        Vector::Subset(Rc::new(RefCell::new(value)), Subsets(Vec::new()))
+    }
+}
+
+impl From<usize> for Subset {
+    fn from(value: usize) -> Self {
+        Subset::Indices(vec![value])
+    }
+}
+
+impl From<Range<usize>> for Subset {
+    fn from(value: Range<usize>) -> Self {
+        Subset::Range(value)
+    }
+}
+
+impl From<Vec<usize>> for Subset {
+    fn from(value: Vec<usize>) -> Self {
+        Subset::Indices(value)
+    }
+}
+
+fn main() {
+    // print out equivalent R code to mock up an interface
+    println!("R 0.v.0 (fake R)");
+
+    let x: Vector<_> = (1..=10).into_iter().collect::<Vec<_>>().into();
+    let y: Vector<_> = vec![10, 9].into();
+    println!("> x <- {}", x);
+    println!("> y <- {}\n> ", y);
+
+    // subset x by some indices
+    let x_subset_1 = x.subset(2..8).subset(vec![1, 2, 4]);
+    println!("> # range and (ordered) vector indexing\n> x[3:8][c(2, 3, 5)]\n{}\n> ", x_subset_1);
+
+    // subset x by some indices
+    let x_subset = x.subset(2..8).subset(vec![4, 2, 2]);
+    println!("> # out-of-order indices\n> x[3:8][c(5, 3, 3)]\n{}\n> ", x_subset);
+
+    // modify
+    x_subset.assign(y);
+    println!("> x[3:8][c(5, 3)] <- y");
+    println!("> x\n{}", x);
+}
+
+#[cfg(test)]
+ mod test {
+    use super::*;
+
+    #[test]
+    fn subset_range() {
+        let x: Vector<_> = (1..=10).into_iter().collect::<Vec<_>>().into();
+        let Vector::Subset(res, _) = x.subset(2..6).materialize();
+        assert_eq!(res.clone().borrow().to_owned(), vec![3, 4, 5, 6])
```

Comment:
What is the purpose of cloning, borrowing then making it owned? My understanding is that `to_owned()` should make an owned object "typically by cloning". In this case I would that that `clone()` alone would make it owned. Is this necessary? Feels like ownership with extrasteps.

Reply:
Just laziness. It felt bad to write, but I wasn't bothered enough to fix it. By the time tests were passing I'd forgotten about the kludge I left behind. Thanks for the spot check, I'll clean it up.

Reply:
Much more sane testing added! Of course it was as easy as `#[derive(PartialEq)]` and now I just compare the results directly.